rootsgetter/rootsgetter.go

```diff
@@ -30,32 +32,44 @@ import (
 
 const logStr = "Roots Getter"
 
+// Storage interface required by Roots Getter.
+type Storage interface {
+	storage.APICallWriter
+	storage.RootsWriter
+}
+
 // Run starts an STH Getter, which periodically queries a Log for its set of acceptable root certificates and stores them.
-func Run(ctx context.Context, lc *client.LogClient, st storage.APICallWriter, l *ctlog.Log, period time.Duration) {
+func Run(ctx context.Context, lc *client.LogClient, st Storage, l *ctlog.Log, period time.Duration) {
 	glog.Infof("%s: %s: started with period %v", l.URL, logStr, period)
 
 	schedule.Every(ctx, period, func(ctx context.Context) {
-		getAndStoreRoots(ctx, lc, st, l)
+		roots, receivedAt, err := getRoots(ctx, lc, st, l)
+		if err != nil {
+			glog.Errorf("%s: %s: %s", l.URL, logStr, err)
+		}
+		if err := st.WriteRoots(ctx, l, roots, receivedAt); err != nil {
+			glog.Errorf("%s: %s: %s", l.URL, logStr, err)
+		}
 	})
 
 	glog.Infof("%s: %s: stopped", l.URL, logStr)
 }
 
-func getAndStoreRoots(ctx context.Context, lc *client.LogClient, st storage.APICallWriter, l *ctlog.Log) {
+func getRoots(ctx context.Context, lc *client.LogClient, st storage.APICallWriter, l *ctlog.Log) ([]*x509.Certificate, time.Time, error) {
 	glog.Infof("%s: %s: getting roots...", l.URL, logStr)
 	roots, httpData, getErr := lc.GetRoots()
-	if getErr != nil {
-		glog.Infof("%s: %s: error getting roots: %s", l.URL, logStr, getErr)
-	} else {
-		glog.Infof("%s: %s: response: %d certificates", l.URL, logStr, len(roots))
-	}
 
 	// Store get-roots API call.
 	apiCall := apicall.New(ct.GetRootsStr, httpData, getErr)
 	glog.Infof("%s: %s: writing API Call...", l.URL, logStr)
 	if err := st.WriteAPICall(ctx, l, apiCall); err != nil {
-		glog.Infof("%s: %s: error writing API Call %s: %s", l.URL, logStr, apiCall, err)
+		return nil, httpData.Timing.End, fmt.Errorf("error writing API Call %s: %s", apiCall, err)
+	}
+
+	if getErr != nil {
+		return nil, httpData.Timing.End, fmt.Errorf("error getting roots: %s", getErr)
```

Comment:
I like the returning errors, it's neater.  However, if *all the things* were to go wrong, and both getting the roots and writing the API call failed, the root-getting error wouldn't be surfaced in the job logs this way, as this function will only return one of these errors, and the caller will only log the one that is returned.

Not sure how to improve it (other than going back to logging right after a bad thing actually happens), but thought it was worth pointing out?

Reply:
Worth pointing out, because when things do go wrong, this is the one tool where we really want to know exactly what it was that was going wrong.  And if the write API call was failing too, there'd be no way to tell that getting the roots also failed (as it won't have been written as part of the API call)

Reply:
`getErr` is stored in `apiCall`, so the error logging should capture both errors.

Reply:
Is that sufficient?

Reply:
Ah I see, sure that should be okay :)